File: ios/framework-objc/liferay-mobile-widgets/login/LoginWidget.m
//
//  LoginPortlet.m
//  liferay-mobile-portlets-sample
//
//  Created by jmWork on 06/06/14.
//  Copyright (c) 2014 Liferay. All rights reserved.
//

#import "LoginWidget.h"
#import "LoginView.h"
#import "LRUserService_v62.h"
#import "BasePortlet+ProgressHUD.h"
#import "LiferayContext.h"
#import "LoginStorage.h"


@interface LoginWidget()

@property (nonatomic, weak) LoginView *loginView;
@property (nonatomic, strong) LRSession *currentSession;

@end


@implementation LoginWidget

- (LoginView *)loginView {
	return (LoginView*)self.portletView;
}

- (void)onCreate {
	self.loginView.username.text = [LiferayContext sharedInstance].currentSession.username;
}

- (void)onCustomAction:(NSString *)actionName fromSender:(UIControl *)sender
{
	if ([sender.restorationIdentifier isEqualToString:@"login-action"]) {
		[self sendLoginWithUsername:self.loginView.username.text
						andPassword:self.loginView.password.text];
	}
}

- (void)sendLoginWithUsername:(NSString *)username andPassword:(NSString *)password
{
	[self showHUDWithMessage:@"Sending login" andDetails:@"Wait few seconds..."];

	[LiferayContext sharedInstance].currentSession = nil;

	self.currentSession = [[LRSession alloc] init:[LiferayContext sharedInstance].server username:username password:password];
	self.currentSession.callback = self;

	LRUserService_v62 *service = [[LRUserService_v62 alloc] initWithSession:self.currentSession];

	NSError *error = nil;
	[service getUserByEmailAddressWithCompanyId:[LiferayContext sharedInstance].companyId emailAddress:username error:&error];

	if (error) {
		[self onFailure:error];
	}
}

- (void)onServerError:(NSError *)error {
	[LiferayContext sharedInstance].currentSession = nil;
	[self hideHUDWithMessage:@"Error in login!" andDetails:nil];
}

- (void)onServerResult:(id)result {
	[LiferayContext sharedInstance].currentSession  = self.currentSession;

	if (self.loginView.remember.on) {
		[LoginStorage storeSession:self.currentSession];
	}
	else {
		[LoginStorage clearStoredSession];
	}

	if ([self.delegate respondsToSelector:@selector(onLogin:withAttributes:)]) {
		id<LoginWidgetDelegate> loginDelegate = (id<LoginWidgetDelegate>) self.delegate;
		[loginDelegate onLogin:self.currentSession withAttributes:result];
	}

	[self hideHUDWithMessage:@"Login completed!" andDetails:nil];
}

@end

File: ios/framework/liferay-ios-widgets/Login/LoginWidget.swift
/**
* Copyright (c) 2000-present Liferay, Inc. All rights reserved.
*
* This library is free software; you can redistribute it and/or modify it under
* the terms of the GNU Lesser General Public License as published by the Free
* Software Foundation; either version 2.1 of the License, or (at your option)
* any later version.
*
* This library is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
* FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
* details.
*/
import UIKit

@objc protocol LoginWidgetDelegate/*: NSObjectProtocol */{

	func onLoginResponse(attributes: Dictionary<String, String>)
	func onLoginError(error: NSError)

	// TODO
	// func onAutologed(session:LRSession)
	// func onCredentialsSaved(session:LRSession)
}


//@objc(LoginWidget)
@IBDesignable class LoginWidget: BaseWidget {
	typealias AuthCall = (String, String, inout NSError) -> (Void)

	enum AuthType {
		case Email, Screenname
	}

	var authType: AuthType = AuthType.Email

	@IBOutlet var delegate: LoginWidgetDelegate?

	override func onCreate() {
		loginView().emailField.text = "test@liferay.com"
	}

	override func onCustomAction(actionName: String, sender: UIControl) {
		if actionName == "login-action" {
			sendLoginWithUsername(loginView().emailField.text, password:loginView().passwordField.text)
		}
	}

	func loginView() -> LoginView {
		return self.widgetView as LoginView
	}

	func sendLoginWithUsername(username:String, password:String) {
		showHUDWithMessage("Sending sign in...", details:"Wait few seconds...")

		let session = LiferayContext.instance.createSession(username, password: password)
		session.callback = self

		let service = LRUserService_v62(session: session)

		if (authType == AuthType.Email) {
			authCallWithEmail(username, password:password, service:service)
		}
		else if (authType == AuthType.Screenname) {
			authCallWithScreenname(username, password:password, service:service)
		}

	}

	func authCallWithEmail(email:String, password:String, service:LRUserService_v62) {
		let companyId: CLongLong = (LiferayContext.instance.companyId as NSNumber).longLongValue

		var outError: NSError?

		service.getUserByEmailAddressWithCompanyId(companyId, emailAddress:email, error:&outError)

		if let error = outError {
			self.onFailure(error)
		}
	}

	func authCallWithScreenname(name:String, password:String, service:LRUserService_v62) {
		let companyId: CLongLong = (LiferayContext.instance.companyId as NSNumber).longLongValue

		var outError: NSError?

		service.getUserByScreenNameWithCompanyId(companyId, screenName:name, error: &outError)

		if let error = outError {
			self.onFailure(error)
		}
	}


	override func onServerError(error: NSError) {
		delegate?.onLoginError(error)
		LiferayContext.instance.clearSession()
		self.hideHUDWithMessage("Error signing in!", details: nil)
	}

	override func onServerResult(result: AnyObject!) {
		delegate?.onLoginResponse(result as Dictionary)
		self.hideHUDWithMessage("Sign in completed", details: nil)
	}

}
